core/test/mpi/base/bindings.cpp

```diff
@@ -0,0 +1,758 @@
+    win.lock_all();
+    win.unlock_all();
+}
+
+
+TYPED_TEST(MpiBindings, CanSendAndRecvValues)
+{
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    auto recv_array = gko::Array<TypeParam>{this->ref};
+    if (my_rank == 0) {
+        auto send_array = std::vector<TypeParam>{1, 2, 3, 4};
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::send<TypeParam>(send_array.data(), 4, rank, 40 + rank,
+                                          comm);
+            }
+        }
+    } else {
+        recv_array = gko::Array<TypeParam>{this->ref, 4};
+        gko::mpi::recv<TypeParam>(recv_array.get_data(), 4, 0, 40 + my_rank,
+                                  comm);
+    }
+    if (my_rank != 0) {
+        auto ref_array = gko::Array<TypeParam>{this->ref, {1, 2, 3, 4}};
+        this->assert_equal_arrays(ref_array, recv_array);
+    }
+}
+
+
+TYPED_TEST(MpiBindings, CanNonBlockingSendAndNonBlockingRecvValues)
+{
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    std::vector<TypeParam> send_array;
+    auto recv_array = gko::Array<TypeParam>{this->ref};
+    TypeParam* data;
+    std::vector<MPI_Request> req1;
+    MPI_Request req2;
+    if (my_rank == 0) {
+        send_array = std::vector<TypeParam>{1, 2, 3, 4};
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                req1.emplace_back(gko::mpi::i_send<TypeParam>(
+                    send_array.data(), 4, rank, 40 + rank, comm));
+            }
+        }
+    } else {
+        recv_array = gko::Array<TypeParam>{this->ref, 4};
+        req2 = std::move(gko::mpi::i_recv<TypeParam>(recv_array.get_data(), 4,
+                                                     0, 40 + my_rank, comm));
+    }
+    if (my_rank == 0) {
+        auto stat1 = gko::mpi::wait_all(req1);
+    } else {
+        auto stat2 = gko::mpi::wait(req2);
+    }
+    if (my_rank != 0) {
+        auto ref_array = gko::Array<TypeParam>{this->ref, {1, 2, 3, 4}};
+        this->assert_equal_arrays(ref_array, recv_array);
+    }
+}
+
+
+TYPED_TEST(MpiBindings, CanPutValuesWithLockAll)
+{
+    using window = gko::mpi::window<TypeParam>;
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    std::vector<TypeParam> data;
+    if (my_rank == 0) {
+        data = std::vector<TypeParam>{1, 2, 3, 4};
+    } else {
+        data = std::vector<TypeParam>{0, 0, 0, 0};
+    }
+    auto win = window(data.data(), 4 * sizeof(TypeParam), comm);
+    win.lock_all();
+    if (my_rank == 0) {
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::put<TypeParam>(data.data(), 4, rank, 0, 4, win);
+                win.flush(rank);
```

Comment:
Mixing member functions with non-member functions seems a bit inconsistent here

Reply:
Here put is equivalent to send and receive, so i think making that member functions of window might not be a good idea.

Reply:
What is the difference between a flush/lock and put/get that motivates one being a member function and one a non-member function? To me, they could be both, but they should be only one consistently.

Reply:
flush/lock are synchronization operations, which put/get are communication operations. To be consistent with other communication operations, I think free functions are more suitable for put/get. Moreover, I don't think it is a good idea to clutter the window class with too many member functions, especially the communication related functions. There are a few others (MPI_Accumulate etc) which are also RDMA and which take window as a parameter, but are more suitable as free functions.

Reply:
MPI_Barrier is a synchronization operation as well, but we use a free function there. So to be consistent, flush/lock/... should be free functions. Personally, I'd slightly prefer a member function-based communication approach, but exclusively free functions would be fine as well.

Reply:
I agree Barrier is currently a free function. I was thinking of moving that into the `communicator` as well.

Reply:
The last change goes in the opposite direction of what I was suggesting - now we have more places where we mix member functions with free functions :smile:

Reply:
Personally, I prefer separation between synchronization operations and communication operations. But if everyone else prefers all functions as free or all functions as member functions, then I am happy to uniformize that. In that case, I would prefer everything as member functions as well.

Maybe @MarcelKoch and @tcojean can chime in ?

Reply:
Just to clarify, with all functions you also mean all the `send`, `recv`, etc. functions, right? In any case, I don't have a strong opinion either way, but I would also support using member functions over free functions.

Reply:
Yes, I mean all the communication functions (send, recv etc.)